Can no longer build godot and godot-nir-static from source together

### Tested versions

- introduced by pr: https://github.com/godotengine/godot/pull/104893

### System information

PopOS 22.04

### Issue description

We use both godot and godot-nir-static repos as submodules within our project.  After updating to godot 4.5 to see if it fixes some other bugs for us, we noticed that we can no longer compile godot against our godot-nir-static submodule.  After bisecting, I narrowed down exactly which file is the culprit: https://github.com/godotengine/godot/pull/104893/files#diff-bb9fd10753b454a0f1381aa36772df32332f7279095791eede1928a4b29db9ed

Changing CPPEXTPATH back to CPPPATH path in this SCsub fixes it.

### Steps to reproduce

Build godot and godot-nir-static from source. godot-nir-static can be found here: https://github.com/godotengine/godot-nir-static.

See that it errors with:

```
In file included from drivers/d3d12/rendering_device_driver_d3d12.cpp:42:
In file included from /home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/src/compiler/spirv/nir_spirv.h:28:
In file included from /home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/src/compiler/nir/nir.h:44:
In file included from /home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/src/compiler/nir_types.h:36:
In file included from /home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/src/compiler/glsl_types.h:42:
In file included from /home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/src/util/glheader.h:72:
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:383:15: error: expected ')'
  383 | typedef void (GL_APIENTRYP PFNGLACTIVETEXTUREPROC) (GLenum texture);
      |               ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:40:33: note: expanded from macro 'GL_APIENTRYP'
Y* 40 | #define GL_APIENTRYP GL_APIENTRnode_dock.cpp
      |                                 ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:383:14: note: to match this '('
  383 | typedef void (GL_APIENTRYP PFNGLACTIVETEXTUREPROC) (GLenum texture);
      |              ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:384:15: error: expected ')'
  384 | typedef void (GL_APIENTRYP PFNGLATTACHSHADERPROC) (GLuint program, GLuint shader);
      |               ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:40:33: note: expanded from macro 'GL_APIENTRYP'
   40 | #define GL_APIENTRYP GL_APIENTRY*
      |                                 ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:384:14: note: to match this '('
  384 | typedef void (GL_APIENTRYP PFNGLATTACHSHADERPROC) (GLuint program, GLuint shader);
      |              ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:384:15: error: typedef redefinition with different types ('void (GLuint, GLuint)' (aka 'void (unsigned int, unsigned int)') vs 'void (GLenum)' (aka 'void (unsigned int)'))
  384 | typedef void (GL_APIENTRYP PFNGLATTACHSHADERPROC) (GLuint program, GLuint shader);
      |               ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:40:22: note: expanded from macro 'GL_APIENTRYP'
   40 | #define GL_APIENTRYP GL_APIENTRY*
      |                      ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/src/util/glheader.h:59:21: note: expanded from macro 'GL_APIENTRY'
   59 | #define GL_APIENTRY GLAPIENTRY
      |                     ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:383:15: note: previous definition is here
  383 | typedef void (GL_APIENTRYP PFNGLACTIVETEXTUREPROC) (GLenum texture);
      |               ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/include/GLES3/gl3.h:40:22: note: expanded from macro 'GL_APIENTRYP'
   40 | #define GL_APIENTRYP GL_APIENTRY*
      |                      ^
/home/tcroc/dev/BlockyBallOT/submodules/godot-nir-static/godot-mesa/src/util/glheader.h:59:21: note: expanded from macro 'GL_APIENTRY'
   59 | #define GL_APIENTRY GLAPIENTRY
      |                     ^
```

Then change CPPEXTPATH to CPPPATH and see that it compile successfully.

### Minimal reproduction project (MRP)

NA

--- a/drivers/d3d12/SCsub
+++ b/drivers/d3d12/SCsub
@@ -162,7 +162,9 @@ else:
         env.Append(CCFLAGS=["-Wno-unknown-pragmas"])
 
 # This is needed since rendering_device_d3d12.cpp needs to include some Mesa internals.
-env_d3d12_rdd.Prepend(CPPEXTPATH=mesa_private_inc_paths)
+# FIXME: Should be CPPEXTPATH, but doing so introduces an include-order bug when combined with
+# godot-nir-static; this necessitates warning macro wrappers. See #106376.
+env_d3d12_rdd.Prepend(CPPPATH=mesa_private_inc_paths)
 # For the same reason as above, the defines must be the same as in the 3rd-party code itself.
 env_d3d12_rdd.Append(CPPDEFINES=extra_defines)
 

--- a/drivers/d3d12/rendering_context_driver_d3d12.cpp
+++ b/drivers/d3d12/rendering_context_driver_d3d12.cpp
@@ -39,7 +39,22 @@
 #include "core/version.h"
 #include "servers/rendering/rendering_device.h"
 
+GODOT_GCC_WARNING_PUSH
+GODOT_GCC_WARNING_IGNORE("-Wmissing-field-initializers")
+GODOT_GCC_WARNING_IGNORE("-Wnon-virtual-dtor")
+GODOT_GCC_WARNING_IGNORE("-Wshadow")
+GODOT_GCC_WARNING_IGNORE("-Wswitch")
+GODOT_CLANG_WARNING_PUSH
+GODOT_CLANG_WARNING_IGNORE("-Wmissing-field-initializers")
+GODOT_CLANG_WARNING_IGNORE("-Wnon-virtual-dtor")
+GODOT_CLANG_WARNING_IGNORE("-Wstring-plus-int")
+GODOT_CLANG_WARNING_IGNORE("-Wswitch")
+GODOT_MSVC_WARNING_PUSH
+
 #include <dxcapi.h>
+
+GODOT_GCC_WARNING_POP
+GODOT_CLANG_WARNING_POP
 
 #if !defined(_MSC_VER)
 #include <guiddef.h>

--- a/drivers/d3d12/rendering_device_driver_d3d12.cpp
+++ b/drivers/d3d12/rendering_device_driver_d3d12.cpp
@@ -40,12 +40,34 @@
 #include "d3d12_godot_nir_bridge.h"
 #include "rendering_context_driver_d3d12.h"
 
+GODOT_GCC_WARNING_PUSH
+GODOT_GCC_WARNING_IGNORE("-Wimplicit-fallthrough")
+GODOT_GCC_WARNING_IGNORE("-Wlogical-not-parentheses")
+GODOT_GCC_WARNING_IGNORE("-Wmissing-field-initializers")
+GODOT_GCC_WARNING_IGNORE("-Wnon-virtual-dtor")
+GODOT_GCC_WARNING_IGNORE("-Wshadow")
+GODOT_GCC_WARNING_IGNORE("-Wswitch")
+GODOT_CLANG_WARNING_PUSH
+GODOT_CLANG_WARNING_IGNORE("-Wimplicit-fallthrough")
+GODOT_CLANG_WARNING_IGNORE("-Wlogical-not-parentheses")
+GODOT_CLANG_WARNING_IGNORE("-Wmissing-field-initializers")
+GODOT_CLANG_WARNING_IGNORE("-Wnon-virtual-dtor")
+GODOT_CLANG_WARNING_IGNORE("-Wstring-plus-int")
+GODOT_CLANG_WARNING_IGNORE("-Wswitch")
+GODOT_MSVC_WARNING_PUSH
+GODOT_MSVC_WARNING_IGNORE(4200) // "nonstandard extension used: zero-sized array in struct/union".
+GODOT_MSVC_WARNING_IGNORE(4806) // "'&': unsafe operation: no value of type 'bool' promoted to type 'uint32_t' can equal the given constant".
+
 #include <nir_spirv.h>
 #include <nir_to_dxil.h>
 #include <spirv_to_dxil.h>
 extern "C" {
 #include <dxil_spirv_nir.h>
 }
+
+GODOT_GCC_WARNING_POP
+GODOT_CLANG_WARNING_POP
+GODOT_MSVC_WARNING_POP
 
 #if !defined(_MSC_VER)
 #include <guiddef.h>

--- a/drivers/d3d12/rendering_shader_container_d3d12.cpp
+++ b/drivers/d3d12/rendering_shader_container_d3d12.cpp
@@ -41,26 +41,10 @@
 #define __REQUIRED_RPCNDR_H_VERSION__ 475
 #endif
 
-#if defined(__GNUC__) && !defined(__clang__)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wnon-virtual-dtor"
-#pragma GCC diagnostic ignored "-Wshadow"
-#pragma GCC diagnostic ignored "-Wswitch"
-#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
-#pragma GCC diagnostic ignored "-Wimplicit-fallthrough"
-#elif defined(__clang__)
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wnon-virtual-dtor"
-#pragma clang diagnostic ignored "-Wstring-plus-int"
-#pragma clang diagnostic ignored "-Wswitch"
-#pragma clang diagnostic ignored "-Wmissing-field-initializers"
-#pragma clang diagnostic ignored "-Wimplicit-fallthrough"
-#endif
-
-#include "d3dx12.h"
+#include <d3dx12.h>
 #include <dxgi1_6.h>
 #define D3D12MA_D3D12_HEADERS_ALREADY_INCLUDED
-#include "D3D12MemAlloc.h"
+#include <D3D12MemAlloc.h>
 
 #include <wrl/client.h>
 
@@ -69,29 +53,34 @@
 #undef MemoryBarrier
 #endif
 
-// No point in fighting warnings in Mesa.
-#if defined(_MSC_VER)
-#pragma warning(push)
-#pragma warning(disable : 4200) // "nonstandard extension used: zero-sized array in struct/union".
-#pragma warning(disable : 4806) // "'&': unsafe operation: no value of type 'bool' promoted to type 'uint32_t' can equal the given constant".
-#endif
+GODOT_GCC_WARNING_PUSH
+GODOT_GCC_WARNING_IGNORE("-Wimplicit-fallthrough")
+GODOT_GCC_WARNING_IGNORE("-Wlogical-not-parentheses")
+GODOT_GCC_WARNING_IGNORE("-Wmissing-field-initializers")
+GODOT_GCC_WARNING_IGNORE("-Wnon-virtual-dtor")
+GODOT_GCC_WARNING_IGNORE("-Wshadow")
+GODOT_GCC_WARNING_IGNORE("-Wswitch")
+GODOT_CLANG_WARNING_PUSH
+GODOT_CLANG_WARNING_IGNORE("-Wimplicit-fallthrough")
+GODOT_CLANG_WARNING_IGNORE("-Wlogical-not-parentheses")
+GODOT_CLANG_WARNING_IGNORE("-Wmissing-field-initializers")
+GODOT_CLANG_WARNING_IGNORE("-Wnon-virtual-dtor")
+GODOT_CLANG_WARNING_IGNORE("-Wstring-plus-int")
+GODOT_CLANG_WARNING_IGNORE("-Wswitch")
+GODOT_MSVC_WARNING_PUSH
+GODOT_MSVC_WARNING_IGNORE(4200) // "nonstandard extension used: zero-sized array in struct/union".
+GODOT_MSVC_WARNING_IGNORE(4806) // "'&': unsafe operation: no value of type 'bool' promoted to type 'uint32_t' can equal the given constant".
 
-#include "nir_spirv.h"
-#include "nir_to_dxil.h"
-#include "spirv_to_dxil.h"
+#include <nir_spirv.h>
+#include <nir_to_dxil.h>
+#include <spirv_to_dxil.h>
 extern "C" {
-#include "dxil_spirv_nir.h"
+#include <dxil_spirv_nir.h>
 }
 
-#if defined(__GNUC__) && !defined(__clang__)
-#pragma GCC diagnostic pop
-#elif defined(__clang__)
-#pragma clang diagnostic pop
-#endif
-
-#if defined(_MSC_VER)
-#pragma warning(pop)
-#endif
+GODOT_GCC_WARNING_POP
+GODOT_CLANG_WARNING_POP
+GODOT_MSVC_WARNING_POP
 
 static D3D12_SHADER_VISIBILITY stages_to_d3d12_visibility(uint32_t p_stages_mask) {
 	switch (p_stages_mask) {

